test/src/Provider/AbstractProviderTest.php

```diff
@@ -185,6 +193,33 @@ public function getHeadersTest()
         $this->assertEquals(['Authorization' => 'Bearer xyz'], $provider->getHeaders($token));
     }
 
+    public function testRandomGeneratorCreatesRandomState()
+    {
+        $xstate = str_repeat('x', 32);
+
+        $generator = m::mock('RandomLib\Generator');
+        $generator->shouldReceive('generateString')->with(32)->times(1)->andReturn($xstate);
+
+        $factory = m::mock('RandomLib\Factory');
+        $factory->shouldReceive('getMediumStrengthGenerator')->times(1)->andReturn($generator);
+
+        $provider = new MockProvider([], ['randomFactory' => $factory]);
+
+        $url = $provider->getAuthorizationUrl();
+
+        parse_str(parse_url($url, PHP_URL_QUERY), $qs);
+
+        $this->assertArrayHasKey('state', $qs);
+        $this->assertSame($xstate, $qs['state']);
+
+        // Same test, but using the non-mock implementation
+        $url = $this->provider->getAuthorizationUrl();
+
+        parse_str(parse_url($url, PHP_URL_QUERY), $qs);
+
+        $this->assertRegExp('/[a-zA-Z0-9\/+]{32}\b/', $qs['state']);
```

Comment:
Not necessary, because the query parameters are built using the `httpBuildQuery` method, which automatically encodes everything. No concerns about decode on the other side, at least nothing we can control. We could provide an explicit character set if that would be an improvement, but I'm not convinced it is.